### ISSUE TITLE
Subpixel additions cause compilation errors for v6.0 users

### ISSUE THREAD

**Issue by sfowlr** [littlevgl/lv_font_conv#32]
The new .subpx value added to the font descriptor doesn't exist in v6.0.x, nor do the LV_FONT_SUBPX_* defines, leading to a build error whenever a new font is generated and included.

I've been able to work around this issue by adding a version check macro around the .subpx definition in the font c file:
```
#if LV_VERSION_CHECK(6,1,0)
    .subpx = LV_FONT_SUBPX_NONE,
#endif
```
lv_table_head.js L73 could be modified to always include this version check on newly generated fonts, providing a permanent backward-compatible solution:
https://github.com/littlevgl/lv_font_conv/blob/26040af79b4f0d4738025dbe45086700bda096b0/lib/writers/lvgl/lv_table_head.js#L73

**Comment by embeddedt** [comment]
Please send a pull request.

**Comment by puzrin** [comment]
@embeddedt i'm not sure it's possible without garbage

@sfowlr Will using `0.3.1` be suitable solution for you if upgrading to 6.1 is not acceptable (but i strongly recommend to upgrade)?

**Comment by embeddedt** [comment]
> i'm not sure it's possible without garbage

I don't see how adding a version check (until 7.0 is released) adds very much "garbage". Could you explain further?

**Comment by puzrin** [comment]
https://github.com/littlevgl/lv_font_conv/blob/master/CHANGELOG.md#040--2019-11-29

BTW changelog says explicit: 0.4+ is for lvgl 6.1+

**Comment by embeddedt** [comment]
The web version uses 0.4 (in order to provide the new features to 6.1 users), so it would still be ideal to add the `#ifdef`.

**Comment by puzrin** [comment]
I'm very sceptical about manual web use instead of automated scripting with semver control :).

But i'm ok in advance with any changes in web & lvgl writer parts, if someone cares about C code verification in different versions.

**Comment by sfowlr** [comment]
Sadly, no option for me to update to 6.1 at the moment, as there's another type-narrowing bug in 6.1 lv_color.c still preventing compilation on some 32-bit targets when using 1 bit-per-pixel. (Looking back through the commit history, people have tried to fix it several times already. I took a quick stab at it myself and couldn't figure it out, either.)

Deleting the .subpx line in my generated font files isn't too big a deal until I can get v6.1 working, but I'm sure I'm not the only one who will run into this issue. I'd put in a PR myself if I wasn't up against a deadline this week.

**Comment by embeddedt** [comment]
> there's another type-narrowing bug in 6.1 lv_color.c still preventing compilation on some 32-bit targets when using 1 bit-per-pixel.

It might be worth opening an issue on the main repo about that, so that one of us knows to take a look.

**Comment by puzrin** [comment]
I'd close this in 2 weeks as stale if no updates.

**Comment by embeddedt** [comment]
To me, that seems a bit hasty, particularly since many people are probably on vacation, and therefore not spending much/any time on this.

Also, the problem with lack of backwards compatibility has been raised by at least 3 different individuals in the last week, so clearly something needs to be done about it.

**Comment by puzrin** [comment]
I don't know anything about other reports. Suggested so only because this issue looked  in my eyes "something endless because nobody cares". I'd like to keep tracker clean but have no intent to upset anyone.

As far as i understand, compatibility "problem" can be resolved in 2 ways:

- Help users to select proper convertor version (that's why semver control invented). Probably, need additional comments in readme.
- Add guards. I'm ok, if that will not require convertor update for 7.0, 8.0,... Personally, such guards do not look aesthetic for such trivial content, but that's only my personal opinion. Only 6.1 become "stable" with font features, and i'd prefer clean font files, without "pre-release" tail of kludges, if that's acceptable for others.

**Comment by kisvegabor** [comment]
I had an idea about it but I forgot to mention that...
`LV_FONT_SUBPX_NONE` is defined as zero. If in case of `LV_FONT_SUBPX_NONE` we don't add `.subpx = LV_FONT_SUBPX_NONE` `.subpx` still will be initialized to zero. 

So I suggest adding `.subpx = LV_FONT_SUBPX_NONE`. What do you think?

**Comment by puzrin** [comment]
https://github.com/littlevgl/lvgl/blob/master/src/lv_font/lv_font_roboto_12.c#L1623

Is it ok to have gap in prop init for all compilers? There were some reports about problems with wrong order. Don't know if gap will cause the same issue or not.

**Comment by kisvegabor** [comment]
Oh, really, it was an issue once. I also don't know if gaps are problematic or not.
We can `subpx` to the last place (not sure it's works better than gaps) but then it tends to be hacky.

I think the `#if` guard would be the best.

**Comment by puzrin** [comment]
> I think the `#if` guard would be the best.

Will it be compatible with 7.0, 8.0 ... releases? I can update JS part if C example provided.

**Comment by embeddedt** [comment]
> Is it ok to have gap in prop init for all compilers?

It looks like the C99 standard requires it: https://stackoverflow.com/a/3374468

**Comment by kisvegabor** [comment]
@embeddedt 
It was a C++ issue: https://github.com/littlevgl/lv_font_conv/issues/25
Maybe the original issue was because he used g++ to compile C files?

cc @Chronic788

**Comment by embeddedt** [comment]
> Maybe the original issue was because he used g++ to compile C files?

Probably. Perhaps we should make it policy in 7.0+ that doing that is not supported, and LittlevGL's files must be compiled using a C compiler. It would be a reasonable solution to the problem, in my opinion, as C++ and C are not the same language.

**Comment by kisvegabor** [comment]
That sounds reasonable. For v6.1 choose with the `#if` guard?

**Comment by embeddedt** [comment]
Yes.

**Comment by kisvegabor** [comment]
@puzrin ?

**Comment by puzrin** [comment]
I'd say, that's not aesthetic :), but i will accept any your choice. If you give me example, how to change generated C code, i will do the rest.

**Comment by kisvegabor** [comment]
Thanks.
The example shown by @sfowlr in the first post is fine:
```c
#if LV_VERSION_CHECK(6,1,0)
    .subpx = LV_FONT_SUBPX_NONE,
#endif
```

**Comment by puzrin** [comment]
@kisvegabor i'm not against feature, but don't like suggested implementation. Because it will break on 7.0.0, 8.0.0 and so on. Can this be done without breaking convertor for new major lvgl releases?

**Comment by puzrin** [comment]
Something like this, but not sure about correctness:

```c
#if LV_VERSION_CHECK(6,0,0)
#if LV_VERSION_CHECK(6,1,0)
    .subpx = LV_FONT_SUBPX_NONE,
#endif
#else
    .subpx = LV_FONT_SUBPX_NONE,
#endif
```

**Comment by puzrin** [comment]
After inspecting how `LV_VERSION_CHECK` works, seems correct equation should be:

```c
#if !LV_VERSION_CHECK(6,0,100)
    .subpx = LV_FONT_SUBPX_NONE,
#endif
```

Please confirm or clarify

**Comment by embeddedt** [comment]
I've personally never needed `LV_VERSION_CHECK`, but according to the comments above it, it looks like `LV_VERSION_CHECK(6,1,0)` is supposed to be true for every version including and above 6.1. Is that not the case?

**Comment by puzrin** [comment]
https://github.com/littlevgl/lvgl/blob/master/src/lv_version.h#L59

As far as i understand, it matches strict major version AND less or equal for lowers. It looks suitable for patches by design. But our case is a bit different, and it finally seems more easy to make direct checks:

```c
#if !(LVGL_VERSION_MAJOR == 6 && LVGL_VERSION_MINOR == 0)
    .subpx = LV_FONT_SUBPX_NONE,
#endif
```

**Comment by puzrin** [comment]
Can anyone confirm/recommend final solution, tested and so on? I'd like to apply patch & release new version.

**Comment by kisvegabor** [comment]
> #if !(LVGL_VERSION_MAJOR == 6 && LVGL_VERSION_MINOR == 0)
    .subpx = LV_FONT_SUBPX_NONE,
#endif

I can confirm it. Really it seems the best solution!

**Comment by puzrin** [comment]
Released 0.4.2.

At least, it still works with 6.1 :)

**Comment by kisvegabor** [comment]
Thank you! 

I've updated the online converter.
